### PR TITLE
Handle negative transaction sequences gracefully

### PR DESCRIPTION
**Pull request by 2opremio**
Also: Create integration test for negative tx sequence submission

Closes https://github.com/stellar/go/issues/3982

### CODE REVIEW

File: services/horizon/internal/integration/protocol18_test.go [hunk 1]
```diff
@@ -759,3 +760,61 @@ func TestLiquidityPoolFailedDepositAndWithdraw(t *testing.T) {
...
+		BaseFee:              txnbuild.MinBaseFee,
+		Timebounds:           txnbuild.NewInfiniteTimeout(),
+		IncrementSequenceNum: false,
+		EnableMuxedAccounts:  true,
+	}
+	tx, err := txnbuild.NewTransaction(txParams)
+	tt.NoError(err)
+	tx, err = tx.Sign(integration.StandaloneNetworkPassphrase, master)
+	tt.NoError(err)
+	txResp, err = itest.Client().SubmitTransaction(tx)
```
Comment (2opremio) [inline services/horizon/internal/integration/protocol18_test.go:817]: This shouldn't fail, but it does

File: services/horizon/internal/txsub/system.go [hunk 2]
```diff
@@ -113,6 +113,16 @@ func (sys *System) Submit(
...
 
+	// TODO: the submission queue system doesn't play well with negative sequence numbers
+	//       which are possible since the XDR uses Int64 (see the uint64(envelope.SeqNum()) casts below
+	//       which will translate sequence -9223372036854775808 into 9223372036854775808).
+	// We could do a bunch of things
+	// 1. Reject transactions with negative values. But this seems excesive since we should just be an intermediary with
+	//    core
+	// 2. Adapt the queuing system to use in64 instead of uint64 ... but that comes with its own problems
+	//    (the nextSequence logic should be handled carefully to avoid overflows)
+	// 3. Avoid the queuing system altogether for negative sequences (which we know are going to fail anyhow).
```
Comment (2opremio) [inline services/horizon/internal/txsub/system.go:124]: @bartekn @tamirms I would appreciate your input here

Comment (2opremio) [reply services/horizon/internal/txsub/system.go:124]: I am thinking that it Core rejects transactions with negative sequence numbers (i.e. they don't even get to be ingested as failed transactions). We could just reject them right away. Thoughts?

Comment (2opremio) [reply services/horizon/internal/txsub/system.go:124]: Actually, the submission system already sends a Badsequence error without contacting core at all, so I think I will just do that

```
	// queue the submission and get the channel that will emit when
	// submission is valid
	seq := sys.SubmissionQueue.Push(sourceAddress, uint64(envelope.SeqNum()))

	// update the submission queue with the source accounts current sequence value
	// which will cause the channel returned by Push() to emit if possible.
	sys.SubmissionQueue.Update(map[string]uint64{
		sourceAddress: sequenceNumber,
	})

	select {
	case err := <-seq:
		if err == sequence.ErrBadSequence {
			// convert the internal only ErrBadSequence into the FailedTransactionError
			err = ErrBadSequence
		}

		if err != nil {
			sys.finish(ctx, hash, response, Result{Err: err})
			return
		}
```

**Review by bartekn — APPROVED** [review]

**Comment by 2opremio** [comment]
I agree about the change being small enough to be included in 2.9.0 , let's merge it!

File: protocols/horizon/main.go [hunk 3]
```diff
@@ -110,8 +112,11 @@ func (a *Account) IncrementSequenceNumber() (int64, error) {
 	if err != nil {
 		return 0, err
 	}
+	if seqNum == int64(math.MaxInt64) {
```
[the remark this answers is not in the document]
Comment (2opremio) [reply protocols/horizon/main.go:115]: (i.e. untyped consts don't seem to exit)
